### PR TITLE
[SPARK-14069][SQL] Improve SparkStatusTracker to also track executor information

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Track executor information like host and port, cache size, running tasks.

## How was this patch tested?

manual test

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rxin

**Comment by cloud-fan** [comment]
I found it's difficulty to write tests for it. As it just collects the informations which are already exposed by the system, is it worth to test them again? cc @rxin

**Comment by rxin** [comment]
It's probably ok.

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/SparkStatusTracker.scala [hunk 1]
```diff
@@ -104,4 +106,34 @@ class SparkStatusTracker private[spark] (sc: SparkContext) {
       }
     }
   }
+
+  /**
+   * Returns a list of all known executors, represented by string with format: "host:port"
+   */
+  def getExecutors(): Array[String] = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/SparkStatusTracker.scala]: It seems kind of arbitrary that `getExecutors` returns `host:port` but not IDs. I think it's better that we make a `SparkExecutorInfo` or something and expose the `host:port` there, along with other things like cache size, `numRunningTasks` etc. Then in the future we can add more things we want to expose without tying ourselves with the `host:port` identifier.

Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkStatusTracker.scala]: If I just want a list of executors, why shouldn't I be able to get them? I think it makes sense to have a more detailed version (maybe replace the following 2), but having a simple one that returns just the list of executors seem to make sense too.

Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkStatusTracker.scala]: The other thing is I don't know if we want to query the scheduler every time we want a list of executors.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/SparkStatusTracker.scala]: we can expose both, but I'd rather call it something more explicit like `getExecutorHostPort` or something. Elsewhere in Spark I would think `getExecutors: Array[String]` returns the executor IDs.

Comment (rxin) [inline core/src/main/scala/org/apache/spark/SparkStatusTracker.scala]: getExecutorList

Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/SparkStatusTracker.scala]: `getExecutorList` LGTM, I'll rename to it.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/SparkStatusTracker.scala]: wait, how is `getExecutorList` different from `getExecutors`? Why not just be more specific what the strings are?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/SparkStatusTracker.scala]: actually let me move this to the main thread so it doesn't get collapsed.

**Comment by andrewor14** [comment]
@rxin I don't understand the distinction between `getExecutors` and `getExecutorList`. Why not just be more explicit about what the strings represent?

The other problem is the keys to the other maps are also expected to be `host:port` but I would normally expect them to be executor IDs. Also host can be confusing because it can be in one of many different formats (e.g. IP addr, all the hostname formats in EC2).

**Comment by andrewor14** [comment]
My proposal:

```
def getExecutorIds(): Array[String]
def getExecutorInfo(executorId: String): Option[SparkExecutorInfo]

private class SparkExecutorInfoImpl(
    val id: String,
    val hostport: String, // host:port
    ...
    val cacheSize: Long,
    val numRunningTasks: Int)
 extends SparkExecutorInfo
```

This is more consistent with the existing status API, where we have things like
```
def getActiveJobIds(): Array[Int]
def getJobInfo(jobId: Int): Option[SparkJobInfo]
```

**Comment by andrewor14** [comment]
screen shot 2016-03-30 at 3 16 08 pm

screen shot 2016-03-30 at 3 16 47 pm

```
scala> df.count()
16/03/30 15:21:58 INFO SparkContext: Starting job: count at <console>:27
...
16/03/30 15:21:58 INFO SparkDeploySchedulerBackend: Launching task 7 on executor id: 3 hostname: 192.168.0.209.
16/03/30 15:21:58 INFO SparkDeploySchedulerBackend: Launching task 8 on executor id: 0 hostname: 192.168.0.209.
16/03/30 15:21:58 INFO SparkDeploySchedulerBackend: Launching task 9 on executor id: 5 hostname: 192.168.0.209.
16/03/30 15:21:58 INFO SparkDeploySchedulerBackend: Launching task 10 on executor id: 1 hostname: 192.168.0.209.
16/03/30 15:21:58 INFO SparkDeploySchedulerBackend: Launching task 11 on executor id: 4 hostname: 192.168.0.209.
16/03/30 15:21:58 INFO SparkDeploySchedulerBackend: Launching task 12 on executor id: 2 hostname: 192.168.0.209.
```

**Comment by rxin** [comment]
Yea that integer id is completely useless to users who want to figure out what to do with their clusters.

**Comment by andrewor14** [comment]
OK, @rxin and I discussed this more offline. Our proposal is:
```
def getExecutorInfos: Seq[SparkExecutorInfo]

private class SparkExecutorInfoImpl(
    val host: String,
    val port: Int,
    ...
    val cacheSize: Long,
    val numRunningTasks: Int)
 extends SparkExecutorInfo
```
Then we don't need to tie us down with the very specific `host:port` format, and we don't have to expose the executor IDs, which are just integers that don't mean much.

**Comment by andrewor14** [comment]
LGTM retest this please

**Comment by andrewor14** [comment]
Merged into master thanks guys.

**Comment by JoshRosen** [comment]
This looks somewhat dodgy to me from a thread-safety perspective since `executorIdToTaskCount` isn't thread-safe and thus we may iterate over it in SparkStatusTracker while also concurrently updating it while processing task updates. I'm going to fix this in a followup but wanted to point it out here since ideally this should not have slipped past code review.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala [hunk 2]
```diff
@@ -90,6 +90,8 @@ private[spark] class TaskSchedulerImpl(
   // Number of tasks running on each executor
   private val executorIdToTaskCount = new HashMap[String, Int]
 
+  def runningTasksByExecutors(): Map[String, Int] = executorIdToTaskCount.toMap
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala]: Adding a `synchronized` here would resolve the thread-safety issue, I think. I'll do this as part of a patch fixing another bug and also touching this line.

**Comment by rxin** [comment]
Why did we merge this when the description says "N/A"?

**Comment by markhamstra** [comment]
@rxin Do you mean the N/A in "How was this patch tested?"  Some guy said that the lack of tests was ok.  https://github.com/apache/spark/pull/11888#issuecomment-200162987

**Comment by rxin** [comment]
Yea but "TODO: tests" and tests: N/A ...

We needed to update the description.

**Pull request closed without merging**